File: singlyLinkedList.js
// Patient Queue

class Patient {
    constructor(name, age, reason_for_visit) {
        this.name = name;
        this.age = age;
        this.reason = reason_for_visit;
    }
}

class Node {
    constructor(patient) {
        this.patient = patient;
        this.next = null;
    }
}

class PatientQueue {
    constructor() {
        this.head = null;
    }

    addPatient(name, age, reason_for_visit) {
        const newPatient = new Patient(name, age, reason_for_visit);
        const newNode = new Node(newPatient);
        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    delPatient(name) {
        if (!this.head) {
            console.log('No patients in the queue.');
            return;
        }
        if (this.head.patient.name === name) {
            this.head = this.head.next;
            return;
        }
        let prev = null;
        let current = this.head;
        while (current) {
            if (current.patient.name === name) {
                prev.next = current.next;
                return;
            }
            prev = current;
            current = current.next;
        }
        console.log(`Patient named ${name} not found in the queue.`);
    }

    displayQueue() {
        if (!this.head) {
            console.log('No patients in the queue.');
            return;
        }
        let current = this.head;
        console.log('\n Patient Queue: \n');
        while (current) {
            let patient = current.patient;
            console.log(`Name: ${patient.name}`);
            console.log(`Age: ${patient.age}`);
            console.log(`Reason For Visit: ${patient.reason}`);
            console.log(`------------------`);
            current = current.next;
        }
    }
}

const patientQueue = new PatientQueue();

patientQueue.addPatient('Jane Doe', 35, 'Chest pains, shortness of breath');
patientQueue.addPatient('John Smith', 45, 'Tripped over dog, swollen knee, pain in leg up to hip');
patientQueue.addPatient('Jim Gordon', 68, 'Low blood pressure, feels faint');

patientQueue.displayQueue();

patientQueue.delPatient('John Smith');

patientQueue.displayQueue();

File: singly_linked_list.py
# Patient Queue

class Patient:
    def __init__(self, name, age, reason_for_visit):
        self.name = name
        self.age = age
        self.reason = reason_for_visit

class Node:
    def __init__(self, patient):
        self.patient = patient
        self.next = None

class PatientQueue:
    def __init__(self):
        self.head = None

    def add_patient(self, name, age, reason_for_visit):
        new_patient = Patient(name, age, reason_for_visit)
        new_node = Node(new_patient)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def delete_patient(self, name):
        if not self.head:
            print('No patients in the queue.')
            return
        if self.head.patient.name == name:
            self.head = self.head.next
            return
        prev = None
        current = self.head
        while current:
            if current.patient.name == name:
                prev.next = current.next
                return
            prev = current
            current = current.next
        print(f'Patient named {name} not found in the queue.')

    def display_queue(self):
        if not self.head:
            print('No patients in the queue.')
            return
        current = self.head
        print('\n Patient Queue: \n')
        while current:
            patient = current.patient
            print(f'Name: {patient.name}')
            print(f'Age: {patient.age}')
            print(f'Reason For Visit: {patient.reason}')
            print('---------------------')
            current = current.next

patient_queue = PatientQueue()

patient_queue.add_patient('Jane Doe', 35, 'Chest pains, shortness of breath')
patient_queue.add_patient('John Smith', 45, 'Tripped over dog, swollen knee, pain in leg up to hip')
patient_queue.add_patient('Jim Gordon', 68, 'Low blood pressure, feels faint')

patient_queue.display_queue()

patient_queue.delete_patient('John Smith')

patient_queue.display_queue()
